### PR TITLE
Restructure Submission to handle new data, updates and deletes

### PR DESCRIPTION
**Pull request by leoraba**
Change Submission structure is required in order to accommodate new records added, edit records and delete records.

Modified endpoints under `submission` tag.:
- GET /submission/{submissionId}
- DELETE /submission/{submissionId}
- DELETE /submission/{submissionId}/{entityName}
- GET /submission/category/{categoryId}
- GET /submission/category/{categoryId}/organization/{organization}
- POST /submission/category/{categoryId}/upload
- POST /submission/category/{categoryId}/commit/{submissionId}

### CODE REVIEW

File: packages/data-model/src/models/submissions.ts [hunk 1]
```diff
@@ -1,22 +1,42 @@
...
+
+export type SubmissionDeleteData = {
+	systemId: string;
+};
+
+export type SubmissionData = {
+	inserts?: Record<string, SubmissionInsertData>;
+	updates?: Record<string, SubmissionUpdateData[]>;
+	deletes?: Record<string, SubmissionDeleteData[]>;
+};
```
Comment (leoraba) [inline packages/data-model/src/models/submissions.ts:35]: This is the new structure of a Submission stored in the database.
Allowing to separate insert data, update and delete changes

File: packages/data-provider/src/services/submissionService.ts [hunk 2]
```diff
@@ -305,29 +312,34 @@ const service = (dependencies: BaseDependencies) => {
...
-						entityName,
-						isValid: false, // By default New Submitted Data is created as invalid until validation process proves otherwise
-						organization: submission.organization,
-						originalSchemaId: submission.dictionaryId,
-						lastValidSchemaId: submission.dictionaryId,
-						systemId: generateIdentifier(entityName, record),
-						createdBy: '', // TODO: get User from auth
-					};
-					return newSubmittedData;
+			if (submission.data?.inserts) {
```
Comment (leoraba) [inline packages/data-provider/src/services/submissionService.ts:322]: So far we only store new data on submission. Future PR's we're going to add `updates` and `deletes` to be included to perform validation

File: packages/data-provider/src/utils/types.ts [hunk 3]
```diff
@@ -206,15 +214,19 @@ export type ActiveSubmissionResponse = {
  * override 'data' object to contain a summary of records
  */
 export type ActiveSubmissionSummaryResponse = Omit<ActiveSubmissionResponse, 'data'> & {
-	data: Record<string, DataActiveSubmissionSummary>;
+	data: {
+		inserts?: Record<string, DataInsertsActiveSubmissionSummary>;
+		updates?: Record<string, DataUpdatesActiveSubmissionSummary>;
+		deletes?: Record<string, DataDeletesActiveSubmissionSummary>;
+	};
```
Comment (leoraba) [inline packages/data-provider/src/utils/types.ts:221]: This is the structure of a submission summary. This is used as a response on 2 GET endpoints

File: packages/data-model/src/models/submissions.ts [hunk 4]
```diff
@@ -1,22 +1,42 @@
...
 
 import { dictionaries } from './dictionaries.js';
 import { dictionaryCategories } from './dictionary_categories.js';
 
 export const submissionStatusEnum = pgEnum('submission_status', ['OPEN', 'VALID', 'INVALID', 'CLOSED', 'COMMITTED']);
 
-export type SubmissionData = {
+export type SubmissionInsertData = {
 	batchName: string;
 	creator: string;
```
Comment (joneubank) [inline packages/data-model/src/models/submissions.ts:17]: Why do we need to track creator here but not on Update and Delete? The user should be tracked in the ActiveSubmission data.

**Review by joneubank — APPROVED** [review]
Good type naming. Proposed an alternative syntax that I would prefer for `reduce<ReturnType>`.

File: packages/data-model/src/models/submissions.ts (hunk 4 above)
Comment (leoraba) [reply packages/data-model/src/models/submissions.ts:17]: The `creator` field here seems to be duplicated as an Active Submission is already by `user`, as you mentioned. Will remove `creator`

File: packages/data-provider/src/utils/submissionUtils.ts [hunk 5]
```diff
@@ -257,17 +259,39 @@ const utils = (dependencies: BaseDependencies) => {
...
-			);
+			const dataInsertsSummary =
+				submission.data?.inserts &&
+				Object.entries(submission.data?.inserts).reduce(
+					(acc, [entityName, entityData]) => {
+						acc[entityName] = { ..._.omit(entityData, 'records'), recordsCount: entityData.records.length };
+						return acc;
+					},
+					{} as Record<string, DataInsertsActiveSubmissionSummary>,
+				);
```
[the remark this answers is not in the document]
Comment (leoraba) [reply packages/data-provider/src/utils/submissionUtils.ts:270]: reduce type updated as suggested

File: packages/data-provider/src/utils/submissionUtils.ts [hunk 6]
```diff
@@ -257,17 +259,39 @@ const utils = (dependencies: BaseDependencies) => {
...
+
+			const dataDeletesSummary =
+				submission.data.deletes &&
+				Object.entries(submission.data?.deletes).reduce(
+					(acc, [entityName, entityData]) => {
+						acc[entityName] = { recordsCount: entityData.length };
+						return acc;
+					},
+					{} as Record<string, DataDeletesActiveSubmissionSummary>,
+				);
```
[the remark this answers is not in the document]
Comment (leoraba) [reply packages/data-provider/src/utils/submissionUtils.ts:290]: Reduce type updated as suggested.

File: packages/data-provider/src/utils/submissionUtils.ts [hunk 7]
```diff
@@ -280,29 +304,29 @@ const utils = (dependencies: BaseDependencies) => {
...
+			const filesDataProcessed: Record<string, SubmissionInsertData> = {};
 			await Promise.all(
 				Object.entries(files).map(async ([entityName, file]) => {
 					const parsedFileData = await tsvToJson(file.path);
 					filesDataProcessed[entityName] = {
 						batchName: file.originalname,
 						creator: userName,
 						records: parsedFileData,
-					} as SubmissionData;
+					} as SubmissionInsertData;
```
[the remark this answers is not in the document]
Comment (leoraba) [reply packages/data-provider/src/utils/submissionUtils.ts]: Converted this `map` function into `reduce` function

[the remark this answers is not in the document]
Comment (leoraba) [reply packages/data-provider/src/utils/submissionUtils.ts]: removed the assestion

**Pull request merged**